### PR TITLE
External calls 2

### PR DESCRIPTION
**Pull request by alexeuler**

### CODE REVIEW

**Comment by alexeuler** [comment]
[Vault external calls validation](https://trello.com/c/xoNfrtvx/272-vault-external-calls-validation)

**Comment by alexeuler** [comment]
[Add raw eth handling for ERC20 vault](https://trello.com/c/IJNl6jet/267-add-raw-eth-handling-for-erc20-vault)

File: contracts/validators/Validator.sol [hunk 1]
```diff
@@ -0,0 +1,15 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/utils/introspection/ERC165.sol";
+import "../interfaces/validators/IValidator.sol";
+import "../interfaces/IProtocolGovernance.sol";
+import "../libraries/ExceptionsLibrary.sol";
```
Comment (MihanixA) [inline contracts/validators/Validator.sol:7]: unused import

**Comment by MihanixA** [comment]
mixing 
```
if (condition) {
    // *checks*
    return;
}
revert();
```
and `require` patterns is hard to read

File: contracts/validators/BaseValidator.sol [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import "../interfaces/IProtocolGovernance.sol";
+import "../libraries/ExceptionsLibrary.sol";
+
+contract BaseValidator is IBaseValidator {
+    IBaseValidator.ValidatorParams internal _validatorParams;
+    IBaseValidator.ValidatorParams internal _stagedValidatorParams;
+    uint256 internal _stagedValidatorParamsTimestamp;
+
+    constructor(IProtocolGovernance protocolGovernance) {
+        _validatorParams = IBaseValidator.ValidatorParams({protocolGovernance: protocolGovernance});
```
[the remark this answers is not in the document]
Comment (alexeuler) [reply contracts/validators/BaseValidator.sol:14]: This is purely for extensibility

**Comment by alexeuler** [comment]
> mixing
> 
> ```
> if (condition) {
>     // *checks*
>     return;
> }
> revert();
> ```
> 
> and `require` patterns is hard to read

What is this referring to?

File: contracts/validators/UniV2Validator.sol [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    function _verifyPath(address[] memory path) private view {
+        IProtocolGovernance protocolGovernance = _validatorParams.protocolGovernance;
+        for (uint256 i = 0; i < path.length - 1; i++) {
+            address token0 = path[i];
+            address token1 = path[i + 1];
+            address pool = factory.getPair(token0, token1);
+            require(token0 != token1, ExceptionsLibrary.INVALID_TOKEN);
+            require(
+                protocolGovernance.hasPermission(pool, PermissionIdsLibrary.ERC20_APPROVE),
```
[the remark this answers is not in the document]
Comment (alexeuler) [reply contracts/validators/UniV2Validator.sol:87]: I think if the pool is approve this implies transfer permission as well

**Review by MihanixA — APPROVED** [review]

**Pull request merged**